### ISSUE TITLE
[LifetimeSafety] Inference/suggest `lifetimebound` with `make_unique` / `std::make_shared` / `unique_ptr(new T)` / `shared_ptr(new T)`

### ISSUE THREAD

**Issue by SidneyCogdill** [llvm/llvm-project#198478]
https://godbolt.org/z/cc81oTzjW

```cpp
#include <memory>

struct Foo {
    Foo(int* p [[clang::lifetimebound]]) : p(p) {}
    int* p;
};

// lifetimebound suggestion triggered, good.
auto f1(int* p) -> Foo
{
    return Foo(p);
}

// lifetimebound suggestion not triggered, bad.
auto f2(int* p) -> std::unique_ptr<Foo> 
{
    return std::make_unique<Foo>(p);
}

// lifetimebound suggestion not triggered, bad.
auto f3(int* p) -> std::shared_ptr<Foo> 
{
    return std::make_shared<Foo>(p);
}

// lifetimebound suggestion not triggered, bad.
auto f4(int* p) -> std::unique_ptr<Foo>
{
    return std::unique_ptr<Foo>(new Foo(p));
}

// lifetimebound suggestion not triggered, bad.
auto f5(int* p) -> std::shared_ptr<Foo>
{
    return std::shared_ptr<Foo>(new Foo(p));
}
```

`-Wlifetime-safety-all -Xclang=-flifetime-safety-inference`
